### PR TITLE
Add support for file transfer on mikrotik router

### PR DESCRIPTION
**Pull request by elavaud**
File transfer support for mikrotik router

### CODE REVIEW

**Comment by elavaud** [comment]
Indeed for the super(). I am backporting old code and haven't spotted that.
I need access to a mikrotik router for the tests. It might take us two weeks. Will keep you updated.

Thanks for the review!

**Comment by ktbyers** [comment]
@elavaud Just pinging you to remind you of this PR. No worries if it is going to take you a bit to get to it.

**Comment by elavaud** [comment]
My apologies, I should have checked that code before the PR. 
It is clearly not working. But we need it so I am going to rework that PR.

**Comment by ktbyers** [comment]
@elavaud Okay, I am just going to mark this as `Blocked` for now. Let me know if it has been updated/fixed.

Thanks.

**Review by Ail1123 — APPROVED** [review]

**Comment by elavaud** [comment]
Hey @ktbyers , who is that Ail1123? This PR is clearly not ready for merging. Did you or the repo got hacked?

**Comment by elavaud** [comment]
Hi @ktbyers,

I finally re-worked that PR.
Here are the scp tests results:

```
(.venv) edouard@UNYC2111LIN205:~/Documents/www/github/netmiko/tests$ py.test -v test_netmiko_scp.py --test_device mikrotik_routeros
============================================================= test session starts =============================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/edouard/Documents/www/github/netmiko/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/edouard/Documents/www/github/netmiko, configfile: setup.cfg
plugins: pylama-7.7.1
collected 12 items                                                                                                                            

test_netmiko_scp.py::test_scp_put PASSED                                                                                                [  8%]
test_netmiko_scp.py::test_remote_space_available PASSED                                                                                 [ 16%]
test_netmiko_scp.py::test_local_space_available PASSED                                                                                  [ 25%]
test_netmiko_scp.py::test_verify_space_available_put PASSED                                                                             [ 33%]
test_netmiko_scp.py::test_remote_file_size PASSED                                                                                       [ 41%]
test_netmiko_scp.py::test_md5_methods SKIPPED (MD5 not supported on this platform)                                                      [ 50%]
test_netmiko_scp.py::test_disconnect PASSED                                                                                             [ 58%]
test_netmiko_scp.py::test_verify_space_available_get PASSED                                                                             [ 66%]
test_netmiko_scp.py::test_scp_get PASSED                                                                                                [ 75%]
test_netmiko_scp.py::test_md5_methods_get SKIPPED (MD5 not supported on this platform)                                                  [ 83%]
test_netmiko_scp.py::test_disconnect_get PASSED                                                                                         [ 91%]
test_netmiko_scp.py::test_file_transfer PASSED                                                                                          [100%]

=========================================================== short test summary info ===========================================================
SKIPPED [1] test_netmiko_scp.py:46: MD5 not supported on this platform
SKIPPED [1] test_netmiko_scp.py:86: MD5 not supported on this platform
================================================== 10 passed, 2 skipped in 72.23s (0:01:12) ===================================================
```

**Comment by elavaud** [comment]
Depends-on #2807 

And re-posting here the test results after the last changes
```
(.venv) edouard@UNYC2111LIN205:~/Documents/www/github/netmiko/tests$ py.test -v test_netmiko_scp.py --test_device mikrotik_routeros
================================================================= test session starts =================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /home/edouard/Documents/www/github/netmiko/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/edouard/Documents/www/github/netmiko, configfile: setup.cfg
plugins: pylama-7.7.1
collected 12 items                                                                                                                                    

test_netmiko_scp.py::test_scp_put PASSED                                                                                                        [  8%]
test_netmiko_scp.py::test_remote_space_available PASSED                                                                                         [ 16%]
test_netmiko_scp.py::test_local_space_available PASSED                                                                                          [ 25%]
test_netmiko_scp.py::test_verify_space_available_put PASSED                                                                                     [ 33%]
test_netmiko_scp.py::test_remote_file_size PASSED                                                                                               [ 41%]
test_netmiko_scp.py::test_md5_methods SKIPPED (MD5 not supported on this platform)                                                              [ 50%]
test_netmiko_scp.py::test_disconnect PASSED                                                                                                     [ 58%]
test_netmiko_scp.py::test_verify_space_available_get PASSED                                                                                     [ 66%]
test_netmiko_scp.py::test_scp_get PASSED                                                                                                        [ 75%]
test_netmiko_scp.py::test_md5_methods_get SKIPPED (MD5 not supported on this platform)                                                          [ 83%]
test_netmiko_scp.py::test_disconnect_get PASSED                                                                                                 [ 91%]
test_netmiko_scp.py::test_file_transfer PASSED                                                                                                  [100%]

=============================================================== short test summary info ===============================================================
SKIPPED [1] test_netmiko_scp.py:46: MD5 not supported on this platform
SKIPPED [1] test_netmiko_scp.py:86: MD5 not supported on this platform
====================================================== 10 passed, 2 skipped in 130.83s (0:02:10) ======================================================
```

**Comment by ktbyers** [comment]
Can you translate this?

```
                remote_cmd = (
                    '{local f FOUND;:if ([:len [/file find name="'
                    f'{self.file_system}/{self.dest_file}"]] > 0) '
                    'do={{:put "YES_$f"}} else={{:put "NOT_$f"}}}'
                )
```

File: tests/test_netmiko_scp.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ def test_scp_get(scp_fixture_get):
 def test_md5_methods_get(scp_fixture_get):
     ssh_conn, scp_transfer = scp_fixture_get
     platform = ssh_conn.device_type
-    if "nokia_sros" in ssh_conn.device_type or "extreme_exos" in ssh_conn.device_type:
+    if ssh_conn.device_type in ["nokia_sros", "extreme_exos", "mikrotik_routeros"]:
```
Comment (ktbyers) [inline tests/test_netmiko_scp.py:85]: Same issue as above (replace with re.search solution).

**Comment by ktbyers** [comment]
What does mikrotik do when you do the following:

```python
            remote_cmd = (
                f'/file print detail where name="{self.file_system}/{remote_file}"'
            )
```

What does Mikrotik show when this works and when this fails?

**Comment by ktbyers** [comment]
I don't really like this as no one in the future will understand it...so I would like a simpler solution:

```
                remote_cmd = (
                    '{local f FOUND;:if ([:len [/file find name="'
                    f'{self.file_system}/{self.dest_file}"]] > 0) '
                    'do={{:put "YES_$f"}} else={{:put "NOT_$f"}}}'
                )
```

I tried to login to a demo mikrotik that I used to have access to, but I guess that doesn't work any more :-(

**Comment by ktbyers** [comment]
Everything looks good on this PR except for the one test issue pertaining to `device_type` and `re_search`

And then the one remote_cmd issue for determining whether the remote file exists.

**Comment by elavaud** [comment]
> What does mikrotik do when you do the following:
> 
> ```python
>             remote_cmd = (
>                 f'/file print detail where name="{self.file_system}/{remote_file}"'
>             )
> ```
> 
> What does Mikrotik show when this works and when this fails?

when it works:
```
[admin@MikroTik] > file print detail where name=flash/test9.txt
 0 name="flash/test9.txt" type=".txt file" size=19 creation-time=jun/02/2022 17:38:39 contents=no logging console 
```
when it doesnt (blank line):
```
[admin@MikroTik] > file print detail where name=flash/test10.txt

```

I will see for the re.seach. Seems easy.

> And then the one remote_cmd issue for determining whether the remote file exists.
I started with a simple solution... I will have a second look on how to make it simpler

**Comment by ktbyers** [comment]
In the doesn't work case, the next thing that comes after this is just the next prompt--correct?

```
[admin@MikroTik] > file print detail where name=flash/test10.txt

[admin@Mikrotok] >
```

So something like the above?

**Comment by elavaud** [comment]
> 

yes

**Comment by ktbyers** [comment]
@elavaud I just made those updates here:

https://github.com/ktbyers/netmiko/pull/2809

It would be good if we could regression test the `test_netmiko_scp.py` tests on that PR (to make sure I didn't mess anything up).

The updated PR will supersede this one.

**Pull request closed without merging**